app/pages/project/instances/instance/tabs/NetworkingTab.tsx

```diff
@@ -256,6 +269,145 @@ export function NetworkingTab() {
+    header: 'Created',
+    cell: (info) => <DateCell value={info.getValue()} />,
+  }),
+]
+
+// Todo: get name from instance, it can be an ID
+// Currently just works with the first NIC
+const AssociatedFirewallRules = () => {
+  const instanceSelector = useInstanceSelector()
+
+  const { data: nicList } = useApiQuery('instanceNetworkInterfaceList', {
+    query: {
+      project: instanceSelector.project,
+      instance: instanceSelector.instance,
+      limit: 10,
+    },
+  })
+
+  const vpcId = nicList?.items[0].vpcId
+
+  const { data: vpc } = useApiQuery(
+    'vpcView',
+    { path: { vpc: vpcId as string } }, // Safe because query is disabled when this is undefined
+    { throwOnError: false, enabled: !!vpcId }
+  )
+
+  const { data: firewallRules } = useApiQuery(
+    'vpcFirewallRulesView',
+    {
+      query: {
+        project: instanceSelector.project,
+        vpc: vpcId as string,
+      },
+    },
+    { enabled: !!vpcId && !!vpc }
+  )
+
+  const { data: subnet } = useApiQuery(
+    'vpcSubnetView',
+    { path: { subnet: nicList?.items[0].subnetId as string } },
+    { throwOnError: false, enabled: nicList && nicList?.items.length > 0 }
+  )
+
+  if (vpc && firewallRules && firewallRules.rules && subnet) {
+    const vpcs = [vpc.name]
+    const subnets = [subnet.name]
+    const ips = nicList && nicList.items.length > 0 ? [nicList?.items[0].ip] : []
+
+    const applicableFirewallRules = firewallRules.rules.filter((rule) =>
+      isRuleApplicableToInstance(rule, instanceSelector.instance, vpcs, subnets, ips)
+    )
+
+    return <FirewallRuleTable data={applicableFirewallRules} />
+  } else {
+    return null
+  }
+}
+
+const FirewallRuleTable = ({ data }: { data: VpcFirewallRule[] }) => {
+  const columns = useMemo(() => {
+    return [...staticColumns]
+  }, [])
+
+  const table = useReactTable({ columns, data })
+  return (
+    <>
+      <Divider className="my-12" />
+
+      <h2 id="firewall-rules-label" className="mb-4 text-mono-sm text-secondary">
+        Associated Firewall Rules
+      </h2>
+      <Table table={table} />
+    </>
+  )
+}
+
+function isRuleApplicableToInstance(
+  rule: VpcFirewallRule,
+  instanceName: string,
+  vpcs: string[],
+  subnets: string[],
+  ips: string[]
+): boolean {
+  for (const target of rule.targets) {
+    switch (target.type) {
+      case 'vpc':
+        if (vpcs.includes(target.value)) {
+          return true
+        }
+        break
+      case 'subnet':
+        if (subnets.includes(target.value)) {
+          return true
+        }
+        break
+      case 'instance':
+        if (instanceName === target.value) {
+          return true
+        }
+        break
+      case 'ip':
+        if (ips.includes(target.value)) {
+          return true
+        }
+        break
+      case 'ip_net':
+        for (const ip of ips) {
+          if (ipRangeCheck(ip, target.value)) {
+            return true
+          }
+        }
+        break
+    }
+  }
+  return false
+}
```

Comment:
Didn't see this function till now. Very fun. Makes me think maybe this can work.